designs/nutanix-datacenterconfig-credentialref.md

```diff
@@ -0,0 +1,51 @@
+# Nutanix Datacenter Config Credential Reference
+
+## Introduction
+
+**Problem:** When a management cluster and a workload cluster are present on two different Nutanix Prism Central instances,
+using same credentials to authenticate to both instances is not possible. To remedy this, we need to associate credentials
+with corresponding Nutanix Datacenter Configs.
+
+## Overview of Solution
+With this feature, a user can specify credentials reference (i.e. Name and Namespace of the secret containing the credentials)
+in the Nutanix Datacenter Config. This info will be fetched and used to authenticate to the Nutanix Prism Central instance
+when creating the workload cluster.
+
+### Solution Details
+Extend the `NutanixDatacenterConfig` CRD to include a new field `credentialRef` storing name, namespace, and kind (similar to how
+`credentialsRef` is defined in the `NutanixCluster` CRD).
+
+```go
+// NutanixDatacenterConfigSpec defines the desired state of NutanixDatacenterConfig.
+type NutanixDatacenterConfigSpec struct {
+    ...
+    // CredentialRef is the reference to the secret that contains the credentials
+    // for the Nutanix Prism Central
+    // +optional
+    CredentialRef *NutanixCredentialReference `json:"credentialRef,omitempty"`
```

Comment:
I think we should document that the controller won't react to changes in the secret itself, and that a new secret needs to be created and the reference updated for the controller to reconcile those changes.

In fact, we should probably recommend that the secrets are created [immutable](https://kubernetes.io/docs/concepts/configuration/secret/#secret-immutable).

Reply:
I don't think secrets need to be immutable. You could re-use the same secret reference for secret rotation, for example. The constraint feels a little unnecessary.

Reply:
Marking secrets as immutable seems a tad bit unnecessary.

Reply:
@thunderboltsid how does this sound?

* We allow secrets to be mutable (and immutable). It’s a user’s choice.
* We watch these referenced secrets, using `OwnerReferences` to link back from secrets to `Cluster`s.
* We won’t reuse the same instance of the secret the user provides for the nutanix provider. EKS-A will have its own secret (user provided) and our controller will create a separate secret for the nutanix provider, making any format changes if necessary.
* The same “eks-a secret” can be referenced in multiple eks-a `Cluster`s.

About the watch part, we can maybe implement it in two phases if timelines are a concern. We can chat offline since that's out of scope for this design.

Reply:
@thunderboltsid could you add those points (reworded however you see fit) to the design?
The rest looks good to me

Reply:
@g-gaston can you please take a look again?